### ISSUE TITLE
Keep running tests on uncaught exceptions

### ISSUE THREAD

**Issue by wperron** [denoland/deno#9213]
While working on #9070 I wrote a test to reproduce the issue, and the uncaught exception stopped the `deno test` process immediately. I think it would be a better user experience if uncaught exceptions were to be caught by the test harness, converted to a test failure and then continue with the rest of the test suite.

The same could also be applied to `Deno.exit` - that's the behavior in Go's `testing` package, coming from that ecosystem, I think that would make sense as well, but that one is debatable.

**Comment by nayeemrmn** [comment]
Do you mean unbound async errors? How can those be attributed to any given test? Which one should fail?

**Comment by wperron** [comment]
I mean if the callback passed to `Deno.test("name", fn() { })` throws, it should simply fail the test instead of crashing and interrupting the whole suite.

For example:
```typescript
Deno.test({
  "testy test",
  async fn() {
    functionToTest(); // <- this throws
  }
});
```

for reference, here's the actual example that I used: https://discord.com/channels/684898665143206084/684911491035430919/801803067108098118

**Comment by nayeemrmn** [comment]
This is already the behaviour. The assert functions work by throwing if the assertion doesn't hold.

**Comment by caspervonb** [comment]
Probably mean unhandled rejection here, and not exception?

**Comment by wperron** [comment]
@nayeemrmn I don't know what to tell you, I guess there are situations where exceptions are thrown in a way that bypass the try/catch in the test harness? maybe then this is more of a bug than a feature request

@caspervonb no I mean Uncaught exception - I get the `error: Uncaught (in promise) Error: test error` error message.

I feel like this is may just be a bug linked to #9070 though 🤔

**Comment by nayeemrmn** [comment]
> I guess there are situations where exceptions are thrown in a way that bypass the try/catch in the test harness? maybe then this is more of a bug than a feature request

Yes, asynchronous errors that haven't been awaited. It's not a bug either, it's a language limitation.

> no I mean Uncaught exception - I get the `error: Uncaught (in promise) Error: test error` error message.

No the correction was right, you're asking for rejections (async errors) to not become exceptions (terminal).

> I feel like this is may just be a bug linked to #9070 though 🤔

No, that is a bug with `pooledMap()`.

**Comment by caspervonb** [comment]
So unhandled rejections are rejections that are not a promise chain, async/await basically just sugars over `.then` and `.catch` after-all.

The only way to handle this would be via the `unhandledrejection` event (which we do not implement at the moment as far as I know).

The problem with trying to catch this is that there is no way to tell which promise chain the rejection originated from.

In practice this means that the test runner may be in the middle of executing say test number 10 when an `unhandledrejection` event caused by code that was executed in test number 1 finally rejects and gets dispatched as an event.

However; I have been experimenting with moving more of the test runner over to Rust in order to run test modules in parallel; I wonder how much of a performance hit we'd take with pre-loading and executing each test in it's own isolate.

Probably a bad idea but would be interesting to do some benchmarks  :thinking:

**Comment by wperron** [comment]
to be fair, I just want to change the behavior of the test harness to give a slightly better experience, `deno run` should totally keep its current behavior IMO. I think you're right, there may not be a way to do this in jsland at all, it might need to be implemented on the Rust side of things.

You bring up some good points about how would that look, and how would tests be parallelized - maybe we should open a discussion for that

**Comment by caspervonb** [comment]
> to be fair, I just want to change the behavior of the test harness to give a slightly better experience

I'm fine with the proposal, just not sure if it can be done with the current JavaScript based test runner 🙃

**Comment by caspervonb** [comment]
From #9528 but thats probably going to be closed so I will continue in here.

> @caspervonb do you think we can fix #9213 at the same time?

Not that simple, as we have the plumbing for window.onunhandledrejection yet (at-least I don't think we do, rusty_v8 callback might have landed).

But I think what we'll can do with this one is that we register a handler for unhandled rejections (be it via JavaScript or in Rust on the isolate) and just count that as a failure and print it as a failure without trying to attribute to any particular test case.

**Comment by bartlomieju** [comment]
This is now supported in `deno test`
